internal/pkg/source/source_test.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"testing"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/analysistest"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+	"golang.org/x/tools/go/ssa"
+)
+
+type testConfig struct {
+	propagatorsPattern string
+	fieldsPattern      string
+	sanitizerPattern   string
+}
+
+func (c *testConfig) IsSanitizer(call *ssa.Call) bool {
+	match, _ := regexp.MatchString(c.sanitizerPattern, call.String())
+	return match
+}
+
+func (c *testConfig) IsPropagator(call *ssa.Call) bool {
+	match, _ := regexp.MatchString(c.propagatorsPattern, call.String())
+	return match
+}
+
+func (c *testConfig) IsSourceFieldAddr(field *ssa.FieldAddr) bool {
+	match, _ := regexp.MatchString(c.fieldsPattern, field.String())
+	return match
+}
+
+type analyzerResult struct {
+	allocations []*ssa.Alloc
+	calls       []*ssa.Call
+	fieldAddr   []*ssa.FieldAddr
+	store       []*ssa.Store
+}
+
+var testAnalyzer = &analysis.Analyzer{
+	Name:       "source",
+	Run:        run,
+	Doc:        "test harness for the logic related to sources",
+	Requires:   []*analysis.Analyzer{buildssa.Analyzer},
+	ResultType: reflect.TypeOf(analyzerResult{}),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	in := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+	var result analyzerResult
+	for _, fn := range in.SrcFuncs {
+		fmt.Printf("Func: %v\n", fn.Name())
+		for _, b := range fn.Blocks {
+			for _, i := range b.Instrs {
+				fmt.Printf("%v : %T\n", i, i)
+
+				switch v := i.(type) {
+				case *ssa.Alloc:
+					result.allocations = append(result.allocations, v)
```

Comment:
I would certainly prefer the approach of the `analysistest`.
However, this would imply creating a new analyzer for each test case, right?
The statement above, assumes that we want to test more than just identifying the source, but also the relationships between the source and its referrers.

Reply:
> However, this would imply creating a new analyzer for each test case, right?
Ahh, that's a fair point.  Depending on how you split things, it might not be every test-case, but could have to be one analyzer for one or more `testdata/` files.

Reply:
The fundamental problem is that ssa instructions do not translate one-to-one to the original Go source code. 
Concretely, if I wanted to use the `analysistest` approach and assert that a given source has the expected graph, I would not be able to do so via comments alone since not all ssa instructions have an equivalent in the original source.
Therefore, a different approach is required. However, I hope that we could improve on the approach suggested by the PR by developing some helper types that reduce the cognitive overhead.